packages/next/next-server/server/next-server.ts

```diff
@@ -1039,24 +1040,71 @@ export default class Server {
       this.dynamicRoutes = this.getDynamicRoutes()
     }
 
+    const rewrites = {
+      beforeFiles,
+      afterFiles,
+      fallback,
+    }
+    const clientHeaders = this.getClientHeaders(rewrites, redirects)
+
     return {
       headers,
       fsRoutes,
-      rewrites: {
-        beforeFiles,
-        afterFiles,
-        fallback,
-      },
+      rewrites,
       redirects,
       catchAllRoute,
       useFileSystemPublicRoutes,
       dynamicRoutes: this.dynamicRoutes,
       basePath: this.nextConfig.basePath,
       pageChecker: this.hasPage.bind(this),
       locales: this.nextConfig.i18n?.locales || [],
+      clientHeaders,
     }
   }
 
+  protected getClientHeaders(
+    rewrites: Router['rewrites'],
+    redirects: Route[]
```

Comment:
Not sure if this covers all the routes that can possibly have `.has` please advise.

Reply:
here is the list of all routes that have `.has`
WL:
```
"source": "/",
"source": "/retention/accessibility",
"source": "/plans",
"source": "/login",
"source": "/signup/:sku",
"source": "/checkout/:page",
```
Legacy
```
"source": "/checkout-form/:path*",
"source": "/set-password",
"source": "/register",
"source": "/plans",
"source": "/login",
"source": "/gift",
"source": "/gift/redeem",
```